roachtest: no volatility for cast sqlsmith/setup=empty/setting=default failed

[(roachtest).sqlsmith/setup=empty/setting=default failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2013639&tab=buildLog) on [master@9eea0be192499fa7d29b5eca5ed14173982b25f0](https://github.com/cockroachdb/cockroach/commits/9eea0be192499fa7d29b5eca5ed14173982b25f0):

```
											)
										),
										(('41 years 10 mons 278 days 15:41:21.695483':::INTERVAL, '':::STRING)),
										(('37 years 127 days 10:37:18.966288':::INTERVAL, e'\x17>!`o':::STRING)),
										(('62 years 777 days 20:47:52.335967':::INTERVAL, e'O.\x04\x1a\x18\x11I':::STRING))
								)
									AS tab_165 (col_284)
								RIGHT JOIN (VALUES ('06:08:31.712812':::TIME)) AS tab_166 (col_285)
									FULL JOIN (
											VALUES
												('01:13:19.05808+13:29:00':::TIMETZ),
												('06:13:39.013972+14:52:00':::TIMETZ),
												('01:10:09.06268-05:06:00':::TIMETZ),
												('22:21:23.701203+10:28:00':::TIMETZ),
												(NULL),
												(
													COALESCE(
														'15:24:24.193463+01:58:00':::TIMETZ,
														'06:47:25.508601+10:50:00':::TIMETZ
													)
												)
										)
											AS tab_167 (col_286) ON true ON NULL,
								(
									VALUES
										(
											COALESCE(3565973807:::OID, 3506951745:::OID),
											ARRAY['-45 years -3 mons -734 days -22:27:05.203906':::INTERVAL,'57 years 9 mons 464 days 02:58:02.027105':::INTERVAL,'-71 years -1 mons -219 days -15:52:43.39243':::INTERVAL,'290 years':::INTERVAL,'46 years 11 mons 740 days 05:39:54.14459':::INTERVAL,'12 years 438 days 11:51:30.856446':::INTERVAL]
										),
										(
											259206075:::OID,
											(NULL::INTERVAL[] || ARRAY['-29 years -6 mons -142 days -16:12:31.810018':::INTERVAL,'58 years 11 mons 529 days 06:07:51.173242':::INTERVAL,'52 years 930 days 17:33:56.673499':::INTERVAL,'1 day':::INTERVAL]::INTERVAL[])::INTERVAL[]
										),
										(
											55508376:::OID,
											ARRAY['10 years 3 mons 845 days 16:45:18.378939':::INTERVAL,'67 years 5 mons 761 days 23:18:58.917765':::INTERVAL]
										),
										(1736996154:::OID, ARRAY[]:::INTERVAL[])
								)
									AS tab_168 (col_287, col_288)
							LIMIT
								1:::INT8
						)
					)
			)
				AS tab_169 (col_290, col_291)
		WHERE
			false
		LIMIT
			39:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=default](https://teamcity.cockroachdb.com/viewLog.html?buildId=2013639&tab=artifacts#/sqlsmith/setup=empty/setting=default)
Related:
- #47541 roachtest: sqlsmith/setup=empty/setting=default failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Ddefault.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

```
06:34:47 test.go:325: test failure: 	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: no volatility for cast unknown::tuple{interval, string}
```

[(roachtest).sqlsmith/setup=empty/setting=default failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2023526&tab=buildLog) on [master@2dd15cd26f95801c678b1034e98a15560dc10baa](https://github.com/cockroachdb/cockroach/commits/2dd15cd26f95801c678b1034e98a15560dc10baa):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=empty/setting=default/run_1
	sqlsmith.go:169,sqlsmith.go:199,test_runner.go:753: error: pq: internal error: no volatility for cast unknown::tuple{string}
		stmt:
		SELECT
			tab_320.col_574 AS col_575
		FROM
			(
				VALUES
					(CASE WHEN (true OR NULL) THEN ('lFRn':::STRING,) ELSE (e'\x16,\x1a\x17edx7':::STRING,) END),
					(NULL),
					(('r5C':::STRING,)),
					(
						(
							SELECT
								CASE WHEN true THEN NULL ELSE NULL END AS col_573
							FROM
								(
									VALUES
										(COALESCE('1976-05-17':::DATE, NULL)),
										(NULL),
										('1982-09-24':::DATE),
										('1981-04-08':::DATE),
										('1979-05-26':::DATE),
										('1991-03-18':::DATE)
								)
									AS tab_319 (col_572)
							WHERE
								true
							LIMIT
								1:::INT8
						)
					),
					(('XQ':::STRING,)),
					((e'\x11':::STRING,))
			)
				AS tab_320 (col_574)
		WHERE
			true
		ORDER BY
			tab_320.col_574 DESC
		LIMIT
			34:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=empty/setting=default](https://teamcity.cockroachdb.com/viewLog.html?buildId=2023526&tab=artifacts#/sqlsmith/setup=empty/setting=default)
Related:
- #47541 roachtest: sqlsmith/setup=empty/setting=default failed  [C-test-failure](https://api.github.com/repos/cockroachdb/cockroach/labels/C-test-failure) [O-roachtest](https://api.github.com/repos/cockroachdb/cockroach/labels/O-roachtest) [O-robot](https://api.github.com/repos/cockroachdb/cockroach/labels/O-robot) [branch-release-20.1](https://api.github.com/repos/cockroachdb/cockroach/labels/branch-release-20.1) [release-blocker](https://api.github.com/repos/cockroachdb/cockroach/labels/release-blocker)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dempty%2Fsetting%3Ddefault.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/opt/memo/testdata/logprops/scalar
+++ b/pkg/sql/opt/memo/testdata/logprops/scalar
@@ -348,3 +348,23 @@ project
                           ├── fd: (3)-->(1,2)
                           ├── prune: (1-3)
                           └── interesting orderings: (+3)
+
+# Regression test for #50258: cast from unknown to tuple.
+expr
+(Values
+  [
+    (Tuple [ (Cast (Null "unknown") "tuple{string}") ] "tuple{string}" )
+  ]
+  [ (Cols [ (NewColumn "a" "tuple{string}") ]) ]
+)
+----
+values
+ ├── columns: a:1(tuple{string})
+ ├── cardinality: [1 - 1]
+ ├── immutable
+ ├── key: ()
+ ├── fd: ()-->(1)
+ ├── prune: (1)
+ └── tuple [type=tuple{string}]
+      └── cast:  [type=tuple{string}]
+           └── null [type=unknown]

--- a/pkg/sql/sem/tree/casts.go
+++ b/pkg/sql/sem/tree/casts.go
@@ -283,6 +283,9 @@ var validCasts = []castInfo{
 	{from: types.StringFamily, to: types.EnumFamily, volatility: VolatilityImmutable},
 	{from: types.EnumFamily, to: types.EnumFamily, volatility: VolatilityImmutable},
 	{from: types.BytesFamily, to: types.EnumFamily, volatility: VolatilityImmutable},
+
+	// Casts to TupleFamily.
+	{from: types.UnknownFamily, to: types.TupleFamily, volatility: VolatilityImmutable},
 }
 
 type castsMapKey struct {

--- a/pkg/sql/sem/tree/casts_test.go
+++ b/pkg/sql/sem/tree/casts_test.go
@@ -130,3 +130,22 @@ func TestCastsVolatilityMatchesPostgres(t *testing.T) {
 		}
 	}
 }
+
+// TestCastsFromUnknown verifies that there is a cast from Unknown defined for
+// all type families.
+func TestCastsFromUnknown(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	for v := range types.Family_name {
+		switch fam := types.Family(v); fam {
+		case types.UnknownFamily, types.AnyFamily:
+			// These type families are exceptions.
+
+		default:
+			cast := lookupCast(types.UnknownFamily, fam)
+			if cast == nil {
+				t.Errorf("cast from Unknown to %s does not exist", fam)
+			}
+		}
+	}
+}

